apps/shell/app.js

```diff
@@ -170,53 +122,85 @@ function default_server_origin(headers){
   return origin;
 }
 
-server.on('upgrade', function upgrade(request, socket, head) {
-  const requestToken = new URLSearchParams(url.parse(request.url).search).get('csrf'),
-        client_origin = request.headers['origin'],
-        server_origin = custom_server_origin(default_server_origin(request.headers));
-
-  var host, dir;
-  [host, dir] = host_and_dir_from_url(request.url);
-
-  if (client_origin &&
-      client_origin.startsWith('http') &&
-      server_origin && client_origin !== server_origin) {
-    socket.write([
-      'HTTP/1.1 401 Unauthorized',
-      'Content-Type: text/html; charset=UTF-8',
-      'Content-Encoding: UTF-8',
-      'Connection: close',
-      'X-OOD-Failure-Reason: invalid origin',
-    ].join('\r\n') + '\r\n\r\n');
-
-    socket.destroy();
+function detect_auth_error(requestToken, client_origin, server_origin, host) {
+  if (host_allowlist.length == 0) {
+    return "No clusters defined.";
+  } else if (client_origin &&
+    client_origin.startsWith('http') &&
+    server_origin && client_origin !== server_origin) {
+    return "Invalid Origin.";
   } else if (!tokens.verify(secret, requestToken)) {
-    socket.write([
-      'HTTP/1.1 401 Unauthorized',
-      'Content-Type: text/html; charset=UTF-8',
-      'Content-Encoding: UTF-8',
-      'Connection: close',
-      'X-OOD-Failure-Reason: bad csrf token',
-    ].join('\r\n') + '\r\n\r\n');
-
-    socket.destroy();
-  } else if (!helpers.hostInAllowList(host_allowlist, host)) { // host not in allowlist
-    socket.write([
-      'HTTP/1.1 401 Unauthorized',
-      'Content-Type: text/html; charset=UTF-8',
-      'Content-Encoding: UTF-8',
-      'Connection: close',
-      'X-OOD-Failure-Reason: host not specified in allowlist or cluster configs',
-    ].join('\r\n') + '\r\n\r\n');
-
-    socket.destroy();
+    return "Bad CSRF Token.";
+  } else if (!helpers.hostInAllowList(host_allowlist, host)) {
+    return "Host not specified in allowlist or cluster configs.";
```

Comment:
apparently there's a 126 byte limit in these messages, or else i'd send a link to the documentation. In the least, I think we can provide the host to give some indication that we know what you're _trying_ to do.

```suggestion
return `${host} not specified in allowlist or cluster configs`;
```